### PR TITLE
SPIKE - change my work experience

### PR DESCRIPTION
**Pull request by dcyoung-dev**
## Context

Allows a Candidate to change their Work Experience after submitting an Application Choice.
Changes to the Work Experience details do not change the submitted Application

## Changes proposed in this pull request

- Modify the ApplicationExperience model to have a polymorphic `exerienceable` relationship rather than being attached directly to an ApplicationForm
- Duplicates the Work Experience record and connects it to the ApplicationChoice when the Application is submitted
  - This is displayed in the Provider Interface
  - This would also need to be used in the `ApplicationPresenter` in the Vendor API

## Guidance to review

This may cause issues elsewhere in the codebase.

It is possible to create an Application, with Work Experience, submit it and then update the Work Experience. 
Viewing the Application as a Provider, the original Work Experience will be visible - without the edits. 

### Running locally

The migrations cause issues because we run our database migrations and then run the data migrations. This PR has a data migration that needs run before a migrations changing a column to not null.

Run `rails db:migrate` this will fail on a `check_constraint` - this is expected in this PR - after this error run `rails data:migrate` and then run `rails db:migrate`.

## Link to Trello card

https://trello.com/c/WfLe897a

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: adr/0024-enabling-work-experience-changes.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+The long term objective of this change, and other that will follow, is to allow the Application Choice to become a snapshot of the Candidate's Application at the time of submission.
+
+### 1. Duplicating records on submission
+
+When a Candidate submits an Application, we will duplicate the Work Experience and Volunteering Experience records, connecting the new duplicates to the Application Choice.
+
+This allows us to enable edits in the current Candidate interface, without affecting the Provider's view of the Application.
+
+The Candidate will always be editing the `ApplicationForm`'s `WorkExperience` and `VolunteeringExperience` records. The Providers will always be viewing the `ApplicationChoice`'s `WorkExperience` and `VolunteeringExperience` records.
```
Comment (elceebee) [inline adr/0024-enabling-work-experience-changes.md:68]: 👍

File: app/services/work_history_with_breaks.rb [hunk 2]
```diff
@@ -1,15 +1,17 @@
 class WorkHistoryWithBreaks
   attr_accessor :application_form, :work_history, :existing_breaks, :unpaid_work, :include_unpaid_experience
 
-  def initialize(application_form, include_unpaid_experience: false)
+  # TODO: This should accept an ApplicationChoice, or an ApplicationForm
+  # Probably rename `application_form` to `experienceable` or similar
```
Comment (elceebee) [inline app/services/work_history_with_breaks.rb:5]: 👍

**Review by elceebee — COMMENTED** [review]
This is looking rather lovely and I think the approach is sound. I haven't tested it locally yet (will do that now), but I do have a couple of questions:
- Should it be behind a feature flag? Is the plan to roll it out immediately or release it next cycle (I know there are always considerations to do with whether or not the change means people applying after the change have an advantage over those who didn't...)
- Do you want to check how this works with carrying over an application to the next cycle? I think it should be fine, but worth testing manually and possibly adding some specs to the `duplicate_application` class.
-

**Comment by dcyoung-dev** [comment]
> This is looking rather lovely and I think the approach is sound. I haven't tested it locally yet (will do that now), but I do have a couple of questions:
> 
> * Should it be behind a feature flag? Is the plan to roll it out immediately or release it next cycle (I know there are always considerations to do with whether or not the change means people applying after the change have an advantage over those who didn't...)
> * Do you want to check how this works with carrying over an application to the next cycle? I think it should be fine, but worth testing manually and possibly adding some specs to the `duplicate_application` class.

Thanks for your comments. We don't plan on merging this PR, only using it as a feasibility check.

We may use a feature flag for this, although we can do the underlying changes without the Candidate noticing. The first commit in this PR makes the section editable to the Candidate which will act like a feature flag. 

Good should on the carried over Applications 🤔 we'll look into this too. Do we carry over the Application Form, the Application Choice or both?

**Comment by dcyoung-dev** [comment]
Closing this PR as the spike is complete.
Please see https://github.com/DFE-Digital/apply-for-teacher-training/pull/9627 for full ADR

**Pull request closed without merging**